### PR TITLE
Refactor ActivityInfoFilter

### PR DESCRIPTION
**Pull request by jdkuki**
## Submitter Checklist:
Resolves brave/brave-browser#5129
- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_rewards/browser/publisher_info_database.cc [hunk 1]
```diff
@@ -325,13 +325,13 @@ PublisherInfoDatabase::GetPublisherInfo(const std::string& publisher_key) {
...
 PublisherInfoDatabase::GetPanelPublisher(
-    const ledger::ActivityInfoFilter& filter) {
+    ledger::ActivityInfoFilterPtr filter) {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
 
   bool initialized = Init();
   DCHECK(initialized);
 
-  if (!initialized || filter.id.empty()) {
+  if (!initialized || filter->id.empty()) {
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/publisher_info_database.cc]: We should check if filter is null here first

File: components/brave_rewards/browser/publisher_info_database.cc [hunk 2]
```diff
@@ -518,44 +518,44 @@ bool PublisherInfoDatabase::GetActivityList(
                       "ON ai.publisher_id = pi.publisher_id "
                       "WHERE 1 = 1";
 
-  if (!filter.id.empty()) {
+  if (!filter->id.empty()) {
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/publisher_info_database.cc]: need filter null check in this if `if (!initialized) {`

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 3]
```diff
@@ -252,30 +252,31 @@ bool SaveActivityInfoOnFileTaskRunner(
 ledger::PublisherInfoList GetActivityListOnFileTaskRunner(
     uint32_t start,
     uint32_t limit,
-    ledger::ActivityInfoFilter filter,
+    ledger::ActivityInfoFilterPtr filter,
     PublisherInfoDatabase* backend) {
   ledger::PublisherInfoList list;
   if (!backend)
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: add null check

File: vendor/bat-native-ledger/include/bat/ledger/public/interfaces/ledger.mojom [hunk 4]
```diff
@@ -95,3 +95,27 @@ struct Balance {
...
+enum EXCLUDE_FILTER {
+  FILTER_ALL = -1,
+  FILTER_DEFAULT = 0,
+  FILTER_EXCLUDED = 1,
+  FILTER_INCLUDED = 2,
+  FILTER_ALL_EXCEPT_EXCLUDED = 3
+};
+
+struct ActivityInfoFilterOrderPair {
+  string propertyName;
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/include/bat/ledger/public/interfaces/ledger.mojom]: this should be `property_name`

File: vendor/brave-ios/tests/ledger_database_test.mm [hunk 5]
```diff
@@ -656,24 +656,24 @@ - (void)testPublishersWithFiltersExcluded
   const auto filter = [[BATActivityInfoFilter alloc] init];
 
   // Filter all - skip exclude filtering
-  filter.excluded = BATExcludeFilterFilterAll;
+  filter.excluded = BATEXCLUDE_FILTERFilterAll;
```
Comment (kylehickinson) [inline vendor/brave-ios/tests/ledger_database_test.mm]: This naming is unfortunate 😅 I will have to fix the formatting for Obj-C generated enum names, or we can stop using all caps snake case for enums in mojo? It seems like it's a bit inconsistent atm: https://github.com/brave/brave-core/pull/2889/files#diff-99c488f19578bfd75399396011e2db1bR99

@NejcZdovc You have any preference?

File: components/brave_rewards/browser/publisher_info_database.cc (hunk 1 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database.cc (hunk 2 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database.cc]: fixed

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 3 above)
Comment (jdkuki) [reply components/brave_rewards/browser/rewards_service_impl.cc]: fixed

File: vendor/bat-native-ledger/include/bat/ledger/public/interfaces/ledger.mojom (hunk 4 above)
Comment (jdkuki) [reply vendor/bat-native-ledger/include/bat/ledger/public/interfaces/ledger.mojom]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 6]
```diff
@@ -360,7 +360,8 @@ TEST_F(PublisherInfoDatabaseTest, GetExcludedList) {
    */
   ledger::ActivityInfoFilter filter;
   filter.excluded = ledger::EXCLUDE_FILTER::FILTER_INCLUDED;
-  success = publisher_info_database_->GetActivityList(0, 0, filter, &pub_list);
+  success = publisher_info_database_->GetActivityList(0, 0,
+      filter.Clone(), &pub_list);
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter.Clone()` should use `std::move(filter)` as `filter` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 7]
```diff
@@ -645,7 +646,7 @@ TEST_F(PublisherInfoDatabaseTest, GetPanelPublisher) {
   filter_4.id = "page.com";
   filter_4.reconcile_stamp = 10;
   ledger::PublisherInfoPtr result =
-      publisher_info_database_->GetPanelPublisher(filter_4);
+      publisher_info_database_->GetPanelPublisher(filter_4.Clone());
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter_4.Clone()` should use `std::move(filter_4)` as `filter_4` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 8]
```diff
@@ -618,15 +619,15 @@ TEST_F(PublisherInfoDatabaseTest, GetPanelPublisher) {
...
+  EXPECT_EQ(publisher_info_database_->GetPanelPublisher(filter_1.Clone()),
             static_cast<ledger::PublisherInfoPtr>(nullptr));
 
   /**
    * Empty table
    */
   ledger::ActivityInfoFilter filter_2;
   filter_2.id = "test";
-  EXPECT_EQ(publisher_info_database_->GetPanelPublisher(filter_2),
+  EXPECT_EQ(publisher_info_database_->GetPanelPublisher(filter_2.Clone()),
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter_2.Clone()` should use `std::move(filter_2)` as `filter_2` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 9]
```diff
@@ -618,15 +619,15 @@ TEST_F(PublisherInfoDatabaseTest, GetPanelPublisher) {
    * Publisher ID is missing
    */
   ledger::ActivityInfoFilter filter_1;
-  EXPECT_EQ(publisher_info_database_->GetPanelPublisher(filter_1),
+  EXPECT_EQ(publisher_info_database_->GetPanelPublisher(filter_1.Clone()),
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter_1.Clone()` should use `std::move(filter_1)` as `filter_1` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 10]
```diff
@@ -835,7 +836,7 @@ TEST_F(PublisherInfoDatabaseTest, GetActivityList) {
   filter_1.excluded = ledger::EXCLUDE_FILTER::FILTER_ALL;
   EXPECT_TRUE(publisher_info_database_->GetActivityList(0,
                                                         0,
-                                                        filter_1,
+                                                        filter_1.Clone(),
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter_1.Clone()` should use `std::move(filter_1)` as `filter_1` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 11]
```diff
@@ -851,7 +852,7 @@ TEST_F(PublisherInfoDatabaseTest, GetActivityList) {
   filter_2.excluded = ledger::EXCLUDE_FILTER::FILTER_ALL;
   EXPECT_TRUE(publisher_info_database_->GetActivityList(0,
                                                         0,
-                                                        filter_2,
+                                                        filter_2.Clone(),
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter_2.Clone()` should use `std::move(filter_2)` as `filter_2` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 12]
```diff
@@ -866,7 +867,7 @@ TEST_F(PublisherInfoDatabaseTest, GetActivityList) {
   filter_3.excluded = ledger::EXCLUDE_FILTER::FILTER_ALL_EXCEPT_EXCLUDED;
   EXPECT_TRUE(publisher_info_database_->GetActivityList(0,
                                                         0,
-                                                        filter_3,
+                                                        filter_3.Clone(),
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter_3.Clone()` should use `std::move(filter_3)` as `filter_3` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 13]
```diff
@@ -885,7 +886,7 @@ TEST_F(PublisherInfoDatabaseTest, GetActivityList) {
   filter_4.excluded = ledger::EXCLUDE_FILTER::FILTER_ALL;
   EXPECT_TRUE(publisher_info_database_->GetActivityList(0,
                                                         0,
-                                                        filter_4,
+                                                        filter_4.Clone(),
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter_4.Clone()` should use `std::move(filter_4)` as `filter_4` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 14]
```diff
@@ -902,7 +903,8 @@ TEST_F(PublisherInfoDatabaseTest, Migrationv3tov4) {
   ledger::PublisherInfoList list;
   ledger::ActivityInfoFilter filter;
   filter.excluded = ledger::EXCLUDE_FILTER::FILTER_ALL;
-  EXPECT_TRUE(publisher_info_database_->GetActivityList(0, 0, filter, &list));
+  EXPECT_TRUE(publisher_info_database_->GetActivityList(0, 0,
+      filter.Clone(), &list));
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter.Clone()` should use `std::move(filter)` as `filter` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 15]
```diff
@@ -923,7 +925,8 @@ TEST_F(PublisherInfoDatabaseTest, Migrationv4tov5) {
...
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter.Clone()` should use `std::move(filter)` as `filter` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 16]
```diff
@@ -946,7 +949,8 @@ TEST_F(PublisherInfoDatabaseTest, Migrationv5tov6) {
...
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter.Clone()` should use `std::move(filter)` as `filter` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 17]
```diff
@@ -987,7 +991,8 @@ TEST_F(PublisherInfoDatabaseTest, Migrationv4tov6) {
...
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter.Clone()` should use `std::move(filter)` as `filter` is no longer used

File: components/brave_rewards/browser/publisher_info_database_unittest.cc [hunk 18]
```diff
@@ -1048,7 +1053,8 @@ TEST_F(PublisherInfoDatabaseTest, DeleteActivityInfo) {
...
```
Comment (tmancey) [inline components/brave_rewards/browser/publisher_info_database_unittest.cc]: `filter.Clone()` should use `std::move(filter)` as `filter` is no longer used

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 19]
```diff
@@ -369,11 +368,11 @@ - (void)listActivityInfoFromStart:(unsigned int)start
...
                        completion:(void (^)(NSArray<BATPublisherInfo *> *))completion
 {
-  const auto cppFilter = filter ? filter.cppObj : ledger::ActivityInfoFilter();
-  if (filter.excluded == BATExcludeFilterFilterExcluded) {
+  const auto cppFilter = filter ? filter.cppObjPtr : ledger::ActivityInfoFilter::New();
+  if (filter.excluded == BATEXCLUDE_FILTERFilterExcluded) {
     completion([BATLedgerDatabase excludedPublishers]);
   } else {
-    ledger->GetActivityInfoList(start, limit, cppFilter, ^(const ledger::PublisherInfoList& list, uint32_t nextRecord) {
+    ledger->GetActivityInfoList(start, limit, cppFilter.Clone(), ^(const ledger::PublisherInfoList& list, uint32_t nextRecord) {
```
[the remark this answers is not in the document]
Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: Not sure how to fix this, bit of a Obj-c++ novice and it appears calling std::move(cppFilter) still somehow invokes the copy constructor. Took the hint from other lines to use .Clone() here

[the remark this answers is not in the document]
Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: `-ios/brave_rewards_ios_framework_arch_shared_library_sources/BATBraveLedger.o
../../brave/vendor/brave-ios/Ledger/BATBraveLedger.mm:388:47: error: call to deleted constructor of 'ledger::ActivityInfoFilterPtr' (aka 'StructPtr<ledger::mojom::ActivityInfoFilter>')
    ledger->GetActivityInfoList(start, limit, std::move(cppFilter),^(const ledger::PublisherInfoList& list, uint32_t nextRecord) {
                                              ^~~~~~~~~~~~~~~~~~~~
../../mojo/public/cpp/bindings/struct_ptr.h:115:28: note: 'StructPtr' has been explicitly marked deleted here
  DISALLOW_COPY_AND_ASSIGN(StructPtr);
                           ^
../../brave/vendor/bat-native-ledger/include/bat/ledger/ledger.h:115:67: note: passing argument to parameter 'filter' here
                                    ledger::ActivityInfoFilterPtr filter,
                                                                  ^
1 error generated.
`

[the remark this answers is not in the document]
Comment (kylehickinson) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: Likely due to the `const` placed when creating `cppFilter` which will force it to copy. 

Try changing:
```objc++
const auto cppFilter = ...
```
to:
```objc++
auto cppFilter = ...
```

[the remark this answers is not in the document]
Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: @kylehickinson did the trick, thanks

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc [hunk 20]
```diff
@@ -108,23 +108,24 @@ void BatPublishers::saveVisit(const std::string& publisher_id,
...
     bool non_verified,
     bool min_visits) {
-  ledger::ActivityInfoFilter filter;
-  filter.id = publisher_id;
-  filter.excluded = excluded;
-  filter.min_duration = min_duration ? getPublisherMinVisitTime() : 0;
-  filter.reconcile_stamp = currentReconcileStamp;
-  filter.non_verified = non_verified;
-  filter.min_visits = min_visits ? getPublisherMinVisits() : 0;
+  ledger::ActivityInfoFilterPtr filter = ledger::ActivityInfoFilter::New();
```
[the remark this answers is not in the document]
Comment (jdkuki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc [hunk 21]
```diff
@@ -108,23 +108,24 @@ void BatPublishers::saveVisit(const std::string& publisher_id,
...
+  filter->id = publisher_id;
+  filter->excluded = excluded;
+  filter->min_duration = min_duration ? getPublisherMinVisitTime() : 0;
+  filter->reconcile_stamp = currentReconcileStamp;
+  filter->non_verified = non_verified;
+  filter->min_visits = min_visits ? getPublisherMinVisits() : 0;
```
[the remark this answers is not in the document]
Comment (jdkuki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc [hunk 22]
```diff
@@ -108,23 +108,24 @@ void BatPublishers::saveVisit(const std::string& publisher_id,
...
-  filter.excluded = excluded;
-  filter.min_duration = min_duration ? getPublisherMinVisitTime() : 0;
-  filter.reconcile_stamp = currentReconcileStamp;
-  filter.non_verified = non_verified;
-  filter.min_visits = min_visits ? getPublisherMinVisits() : 0;
+  ledger::ActivityInfoFilterPtr filter = ledger::ActivityInfoFilter::New();
+  filter->id = publisher_id;
+  filter->excluded = excluded;
+  filter->min_duration = min_duration ? getPublisherMinVisitTime() : 0;
+  filter->reconcile_stamp = currentReconcileStamp;
```
[the remark this answers is not in the document]
Comment (jdkuki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/bat_publishers.cc]: fixed

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 23]
```diff
@@ -543,21 +544,22 @@ void RewardsServiceImpl::GetContentSiteList(
     uint32_t min_visits,
     bool fetch_excluded,
     const GetContentSiteListCallback& callback) {
-  ledger::ActivityInfoFilter filter;
-  filter.min_duration = min_visit_time;
-  filter.order_by.push_back(std::pair<std::string, bool>("ai.percent", false));
-  filter.reconcile_stamp = reconcile_stamp;
-  filter.excluded = fetch_excluded
+  ledger::ActivityInfoFilterPtr filter = ledger::ActivityInfoFilter::New();
```
[the remark this answers is not in the document]
Comment (jdkuki) [reply components/brave_rewards/browser/rewards_service_impl.cc]: fixed

File: vendor/brave-ios/Ledger/BATBraveLedger.mm (hunk 19 above)
[the remark this answers is not in the document]
Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: @tmancey fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 6 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 8 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 9 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 11 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 12 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 13 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 14 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 18 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 15 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 16 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: components/brave_rewards/browser/publisher_info_database_unittest.cc (hunk 17 above)
Comment (jdkuki) [reply components/brave_rewards/browser/publisher_info_database_unittest.cc]: fixed

File: vendor/brave-ios/tests/ledger_database_test.mm (hunk 5 above)
Comment (tmancey) [reply vendor/brave-ios/tests/ledger_database_test.mm]: @kylehickinson I have spoken with @NejcZdovc and for now this is ok and we will raise a ticket for consistency

Comment (kylehickinson) [reply vendor/brave-ios/tests/ledger_database_test.mm]: Gotcha, I will push a mojo gen fix to this branch that refactors the all caps snake case then

**Pull request merged**